flexmeasures/data/models/generic_assets.py

```diff
@@ -164,6 +167,127 @@ def __repr__(self):
             self.generic_asset_type.name,
         )
 
+    def validate_sensors_to_show(
+        self,
+    ) -> list[dict[str, str | None | "Sensor" | list["Sensor"]]]:  # noqa: F821
+        """
+        Validate and transform the 'sensors_to_show' attribute into the latest format for use in graph-making code.
+
+        This function ensures that the 'sensors_to_show' attribute:
+        1. Follows the latest format, even if the data in the database uses an older format.
+        2. Contains only sensors that the user has access to (based on the current asset, account, or public availability).
+
+        Steps:
+        - The function deserializes the 'sensors_to_show' data from the database, ensuring that older formats are parsed correctly.
+        - It checks if each sensor is accessible by the user and filters out any unauthorized sensors.
+        - The sensor structure is rebuilt according to the latest format, which allows for grouping sensors and adding optional titles.
+
+        Details on format:
+        - The 'sensors_to_show' attribute is defined as a list of sensor IDs or nested lists of sensor IDs (to indicate grouping).
+        - Titles may be associated with rows of sensors. If no title is provided, `{"title": None}` will be assigned.
+        - Nested lists of sensors indicate that they should be shown together (e.g., layered in the same chart).
+
+        Example inputs:
+        1. Simple list of sensors, where sensors 42 and 44 are grouped:
+            sensors_to_show = [40, 35, 41, [42, 44], 43, 45]
+
+        2. List with titles and sensor groupings:
+            sensors_to_show = [
+                {"title": "Title 1", "sensor": 40},
+                {"title": "Title 2", "sensors": [41, 42]},
+                [43, 44], 45, 46
+            ]
+
+        Returned structure:
+        - The function returns a list of dictionaries, with each dictionary containing either a 'sensor' (for individual sensors) or 'sensors' (for groups of sensors), and an optional 'title'.
+        - Example output:
+            [
+                {"title": "Title 1", "sensor": <Sensor object for sensor 40>},
+                {"title": "Title 2", "sensors": [<Sensor object for sensor 41>, <Sensor object for sensor 42>]},
+                {"title": None, "sensors": [<Sensor object for sensor 43>, <Sensor object for sensor 44>]},
+                {"title": None, "sensor": <Sensor object for sensor 45>},
+                {"title": None, "sensor": <Sensor object for sensor 46>}
+            ]
+
+        If the 'sensors_to_show' attribute is missing, the function defaults to showing two of the asset's sensors, grouped together if they share the same unit, or separately if not.
+
+        Unauthorized sensors are filtered out, and a warning is logged. Only sensors the user has permission to access are included in the final result.
+        """
+
+        self.sensors_to_show = (
+            self.attributes.get("sensors_to_show", [])
```

Comment:
Let's leave this out  - we should not rely anymore on this being in self.attributes. We copied existing values over.